--- a/fs/squashfs/inode.c.orig	2007-09-23 09:31:03.567520222 +0200
+++ b/fs/squashfs/inode.c	2007-09-23 09:32:27.061781932 +0200
@@ -32,6 +32,7 @@
 #include <linux/vmalloc.h>
 #include <linux/smp_lock.h>
 
+#include <linux/exportfs.h>
 #include "squashfs.h"
 
 static void vfs_read_inode(struct inode *i);
@@ -2307,7 +2308,7 @@
 	squashfs_inode_cachep = kmem_cache_create("squashfs_inode_cache",
 	     sizeof(struct squashfs_inode_info),
 	     0, SLAB_HWCACHE_ALIGN|SLAB_RECLAIM_ACCOUNT,
-	     init_once, NULL);
+	     init_once);
 	if (squashfs_inode_cachep == NULL)
 		return -ENOMEM;
 	return 0;
