diff -r -u linux-2.6.13.4-orig/include/asm-i386/param.h linux-2.6.13.4/include/asm-i386/param.h
--- linux-2.6.13.4-orig/include/asm-i386/param.h	2005-10-10 11:54:29.000000000 -0700
+++ linux-2.6.13.4/include/asm-i386/param.h	2005-10-15 23:08:15.000000000 -0700
@@ -20,6 +20,6 @@
 #endif
 
 #define MAXHOSTNAMELEN	64	/* max length of hostname */
-#define COMMAND_LINE_SIZE 256
+#define COMMAND_LINE_SIZE 512
 
 #endif
diff -r -u linux-2.6.13.4-orig/include/asm-i386/setup.h linux-2.6.13.4/include/asm-i386/setup.h
--- linux-2.6.13.4-orig/include/asm-i386/setup.h	2005-10-10 11:54:29.000000000 -0700
+++ linux-2.6.13.4/include/asm-i386/setup.h	2005-10-15 23:08:27.000000000 -0700
@@ -17,7 +17,7 @@
 #define MAX_NONPAE_PFN	(1 << 20)
 
 #define PARAM_SIZE 4096
-#define COMMAND_LINE_SIZE 256
+#define COMMAND_LINE_SIZE 512
 
 #define OLD_CL_MAGIC_ADDR	0x90020
 #define OLD_CL_MAGIC		0xA33F
