--- linux-2.6.16-rc6/drivers/scsi/iteraid.c	2006-03-18 11:51:53.000000000 +0100
+++ linux-2.6.16-rc5/drivers/scsi/iteraid.c	2006-03-11 00:52:10.000000000 +0100
@@ -4728,7 +4728,7 @@
 /*
  * Find and initialize any cards.
  */
-static int iteraid_detect(Scsi_Host_Template * tpnt)
+static int iteraid_detect(struct scsi_host_template * tpnt)
 {
 	u8 i;
 	u8 j;
@@ -5130,7 +5130,7 @@
 	return 0;
 }
 
-static Scsi_Host_Template driver_template = {
+static struct scsi_host_template driver_template = {
 	.proc_name = 			"IT8212",
 	.proc_info =			iteraid_proc_info,
 	.name =				"ITE RAIDExpress133",
