--- linux-2.6.16-rc6/drivers/char/speakup/speakup.c	2006-03-18 11:51:51.000000000 +0100
+++ linux-2.6.16-rc5/drivers/char/speakup/speakup.c	2006-03-10 22:00:37.000000000 +0100
@@ -479,7 +479,8 @@
       continue;
     }
     count = sel_buffer_lth - pasted;
-    count = MIN(count, tty->ldisc.receive_room(tty));
+    // count = MIN(count, tty->ldisc.receive_room(tty));
+    count = MIN(count, tty->receive_room);
     tty->ldisc.receive_buf(tty, sel_buffer + pasted, 0, count);
     pasted += count;
   }
