--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -1929,6 +1929,18 @@
 		}
 	}
 
+	/*
+	 * Block write, discard, erase, flush, zone reset commands going to a
+	 * read-only device. We do this because kernel drivers often lack
+	 * necessary checks and send these commands to read-only block devices.
+	 */
+	if (unlikely((bio_op(bio) != REQ_OP_READ)
+			&& bdev_read_only(bio->bi_bdev))) {
+		pr_warn("unexpected write command to %s blocked\n",
+			bdevname(bio->bi_bdev, b));
+		goto end_io;
+	}
+
 	switch (bio_op(bio)) {
 	case REQ_OP_DISCARD:
 		if (!blk_queue_discard(q))
