--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -2178,6 +2178,20 @@ generic_make_request_checks(struct bio *bio)
 		}
 	}
 
+	/*
+	 * Block write, discard, erase, flush, zone reset commands going to a
+	 * read-only device. We do this because kernel drivers often lack
+	 * necessary checks and send these commands to read-only block devices.
+	 */
+	struct block_device *bdev = bdget(disk_devt((bio)->bi_disk));
+	if (likely(bdev != NULL) &&
+	    unlikely((bio_op(bio) != REQ_OP_READ) &&
+	    bdev_read_only(bdev))) {
+		pr_warn("unexpected write command to %s blocked\n",
+			bio_devname(bio, b));
+		goto end_io;
+	}
+
 	switch (bio_op(bio)) {
 	case REQ_OP_DISCARD:
 		if (!blk_queue_discard(q))
