From dcdc6b8470e128296e47162e0c04b4160ef8daa2 Mon Sep 17 00:00:00 2001
From: Michael Prokop <mika@grml.org>
Date: Mon, 22 Nov 2010 18:28:17 +0000
Subject: 4200_vesa_fallback_on_fail.patch

---
 arch/x86/boot/video.c |   12 ++++++++++--
 1 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/arch/x86/boot/video.c b/arch/x86/boot/video.c
index 43eda28..380a520 100644
--- a/arch/x86/boot/video.c
+++ b/arch/x86/boot/video.c
@@ -315,6 +315,7 @@ static void restore_screen(void)
 void set_video(void)
 {
 	u16 mode = boot_params.hdr.vid_mode;
+	int vesa_failed = 0;
 
 	RESET_HEAP();
 
@@ -329,8 +330,15 @@ void set_video(void)
 		if (!set_mode(mode))
 			break;
 
-		printf("Undefined video mode number: %x\n", mode);
-		mode = ASK_VGA;
+		if (!vesa_failed) {
+			/* fallback is vesa 640x480 16bit */
+			printf("Could not set mode: %x, trying 640x480-16\n", mode);
+			mode = 0x311;
+			vesa_failed = 1;
+		} else {
+			printf("Undefined video mode number: %x, falling back to text mode\n", mode);
+			mode = VIDEO_80x25;
+		}
 	}
 	boot_params.hdr.vid_mode = mode;
 	vesa_store_edid();
-- 
1.7.2.3

