--- linux-2.6.11-gentoo-r8/drivers/Makefile.orig	2005-05-17 00:07:13.000000000 +0100
+++ linux-2.6.11-gentoo-r8/drivers/Makefile	2005-05-17 00:08:28.000000000 +0100
@@ -23,9 +23,9 @@ obj-$(CONFIG_FB_INTEL)          += video
 # serial drivers start registering their serio ports
 obj-$(CONFIG_SERIO)		+= input/serio/
 obj-y				+= serial/
+obj-$(CONFIG_PARPORT)		+= parport/
 # load keyboard early so speakup can be quieted
 obj-$(CONFIG_INPUT)		+= input/
-obj-$(CONFIG_PARPORT)		+= parport/
 obj-y				+= base/ block/ misc/ net/ media/
 obj-$(CONFIG_NUBUS)		+= nubus/
 obj-$(CONFIG_ATM)		+= atm/
