From beaa24ee0ead413ce638e818623e2afea906e033 Mon Sep 17 00:00:00 2001
From: Michael Prokop <mika@grml.org>
Date: Sat, 8 Mar 2008 15:30:27 +0000
Subject: [PATCH 4/4] Apply x86/5000_grml-kernelversion.patch


Signed-off-by: Michael Prokop <mika@grml.org>

diff --git a/Makefile b/Makefile
index ae78a31..a06cd02 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 25
-EXTRAVERSION = -rc4
+EXTRAVERSION = -grml-rc4
 NAME = Funky Weasel is Jiggy wit it
 
 # *DOCUMENTATION*
-- 
1.5.4.3

