From 9a0b73df09b88d5f264df5df6c63957834eeb59e Mon Sep 17 00:00:00 2001
From: Michael Prokop <mika@grml.org>
Date: Tue, 25 Oct 2011 15:55:19 +0200
Subject: [PATCH] add grml as extraversion

---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 07bc925..be6db82 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 3
 PATCHLEVEL = 1
 SUBLEVEL = 0
-EXTRAVERSION =
+EXTRAVERSION = -grml
 NAME = "Divemaster Edition"
 
 # *DOCUMENTATION*
-- 
1.7.7

